test/test_arhp.rb

```diff
@@ -55,6 +55,22 @@ def test_should_insert_on_correct_database
     assert_action_uses_correct_database(:insert, "insert into tests values(NULL, 'foo')")
   end
 
+  def test_models_with_matching_hosts_and_non_matching_databases_should_share_a_connection
+    simulate_rails_app_active_record_railties
+    assert_equal(Test1.connection.raw_connection, Test1Shard.connection.raw_connection)
+  end
+
+  def test_models_with_matching_hosts_and_non_matching_databases_do_not_mix_up_underlying_database
+    simulate_rails_app_active_record_railties
+
+    # ActiveRecord 6.0 introduced a change that surfaced a problematic code
+    # path in active_record_host_pool: Clearing the cache across connection
+    # handlers can cause the shared host pool to change the underlying
+    # database out from under an in-progress operation.
+    # See lib/active_record_host_pool/connection_handling.rb for more
+    ActiveRecord::Base.cache { Test1Shard.create! }
```

Comment:
Could we add a test here to verify that it also fails? E.g.

```ruby
# before ResetActiveDatabaseAfterClearingCache is prepended
exception = assert_raises(ActiveRecord::StatementInvalid) do
  ActiveRecord::Base.cache { Test1Shard.create! }
end

assert_equal("Mysql2::Error: Table 'arhp_test_2.test1_shards' doesn't exist", exception.message)
```

That would verify that our simulation works as expected and let us know if the patch is no longer needed in newer Rails versions.

Reply:
Great idea. I added a new test that removes the patch, exercises the code, and re-adds the patch. 

That code is here: https://github.com/zendesk/active_record_host_pool/pull/61/files#diff-1f28086f9444c322d177103d79827deeR64-R77